src/uri_util.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+
+use http::uri;
+use std::path;
+use std::vec;
+
+fn _normalize_path(path_str: &str) -> String {
+  let path = path::Path::new(path_str);
+  let mut normalized_path_list: vec::Vec<&str> = vec::Vec::new();
+  for component in path.components() {
+    if component == path::Component::ParentDir {
+      normalized_path_list.pop();
+    } else if component != path::Component::RootDir
+      && component != path::Component::CurDir
+    {
+      normalized_path_list.push(component.as_os_str().to_str().unwrap());
+    }
+  }
+
+  format!("/{}", normalized_path_list.join("/"))
+}
+
+pub fn _uri_join(
+  base: uri::Uri,
+  path: &str,
+) -> Result<uri::Uri, uri::InvalidUriParts> {
+  let base_parts = base.into_parts();
+  let base_path_and_query = base_parts.path_and_query.unwrap();
+  let base_path = base_path_and_query.path();
+  // TODO(zino): Should consider query parameter as well.
+  let joined_path = _normalize_path(&(String::from(base_path) + "/" + path));
+  let joined_path_and_query =
+    uri::PathAndQuery::from_shared(joined_path.into()).unwrap();
+  let mut joined_parts = uri::Parts::default();
+  joined_parts.scheme = base_parts.scheme;
+  joined_parts.authority = base_parts.authority;
+  joined_parts.path_and_query = Some(joined_path_and_query);
+  uri::Uri::from_parts(joined_parts)
+}
+
+#[test]
+fn test_uri_join_basic() {
+  let base = "https://example.com".parse::<uri::Uri>().unwrap();
+  let expected = "https://example.com/hello/world"
+    .parse::<uri::Uri>()
+    .unwrap();
+  assert_eq!(expected, _uri_join(base, "hello/world").unwrap());
+}
+
+#[test]
+fn test_uri_join_with_dot() {
+  let base = "https://example.com".parse::<uri::Uri>().unwrap();
+  let expected = "https://example.com/hello/world"
+    .parse::<uri::Uri>()
+    .unwrap();
+  assert_eq!(expected, _uri_join(base, "./hello/./world").unwrap());
+}
+
+#[test]
+fn test_uri_join_with_dotdot() {
+  let base = "https://example.com".parse::<uri::Uri>().unwrap();
+  let expected = "https://example.com/world".parse::<uri::Uri>().unwrap();
+  assert_eq!(expected, _uri_join(base, "hello/../world").unwrap());
```

Comment:
Can we get a more complex example with more than one ".." ?
Also a test where there are multiple slashes "hello///world"
And also a test where the base URL ends in a slash. EG "https://example.com/foo/bar/"

Reply:
Sure, I'll do it soon. Thank you for review!